examples/tabular_q.py

```diff
@@ -0,0 +1,169 @@
+        hashed = hash_state_action(ob, act, step)
+        if hashed not in q_table:
+            q_table[hashed] = 0
+        max_q = max(q_table[hashed], max_q)
+    return max_q
+
+
+def train(q_table, env):
+    # Buffer an old version of q table to inspect training progress
+    prev_q = {}
+
+    for i in range(FLAGS.episodes):
+        current_length = 0
+        obs = env.reset()
+        while current_length < FLAGS.episode_length:
+            # Run Epsilon greedy policy.
+            a = select_action(q_table, obs, current_length, FLAGS.epsilon)
+            hashed = hash_state_action(obs, a, current_length)
+            if hashed not in q_table:
+                q_table[hashed] = 0
+            obs, reward, done, info = env.step(env.action_space.flags.index(a))
+            current_length += 1
+            # Get max q at new state.
+            target = reward + FLAGS.discount * get_max_q_value(
+                q_table, obs, current_length
+            )
+            # Update Q value at current state action pair.
+            q_table[hashed] = (
+                FLAGS.learning_rate * target
+                + (1 - FLAGS.learning_rate) * q_table[hashed]
+            )
+
+        if i % 50 == 0:
+            print(f"Running episode {i}")
+
+            def compare_qs(q_old, q_new):
+                diff = [q_new[k] - v for k, v in q_old.items()]
+                return sum(diff) / len(diff) if diff else "NaN"
+
+            difference = compare_qs(prev_q, q_table)
+            print(
+                f"Newly added Q entries {len(q_table)-len(prev_q)}, averaged diff {difference}"
+            )
+            prev_q = q_table.copy()
+
+
+def setup_env():
+    FLAGS.observation = "Autophase"
+    FLAGS.reward = "IrInstructionCount"
+    FLAGS.benchmark = "cBench-v0/dijkstra"
+    FLAGS.env = "llvm-ic-v0"
+
+
+def main(argv):
+    setup_env()
+    # Train a Q table.
+    q_table = {}
+    env = get_env()
+    try:
+        started = time()
+        train(q_table, env)
+        print(f"Time spent training {humanize.naturaldelta(time() - started)}")
+        # Rollout based on the Max-Q policy.
+        ob = env.reset()
+        # Roll out one episode and report the resulting policy.
+        action_seq, rewards = [], []
+        for i in range(FLAGS.episode_length):
+            a = select_action(q_table, ob, i)
+            action_seq.append(a)
+            ob, reward, done, info = env.step(env.action_space.flags.index(a))
+            rewards.append(reward)
+        print(
+            "Resulting sequence: ", ",".join(action_seq), f"total reward {sum(rewards)}"
+        )
+        # Rollout the best episode from brute force solution.
+        best = ["-gvn-hoist", "-newgvn", "-instcombine", "-mem2reg", "-simplifycfg"]
+        rewards = []
+        env.reset()
+        for a in best:
+            _, r, _, _ = env.step(env.action_space.flags.index(a))
+            rewards.append(r)
+        print(f"Best achievable result {sum(rewards)}")
```

Comment:
```suggestion
```

I don't think we should include a hardcoded policy. I understand why it's useful to compare against a global optimum, but this is too fragile, and will break if a user tries a different benchmark/reward etc. I think we leave it as an exercise to the reader to figure out how good the solution is :)